driver/src/main/java/org/dcache/SapphireDriver.java

```diff
@@ -82,7 +97,178 @@ public void flush(Iterable<FlushRequest> requests)
     @Override
     public void stage(Iterable<StageRequest> requests)
     {
-        throw new UnsupportedOperationException("Not implemented");
+        _log.debug("Stage triggered");
+
+        for(StageRequest stageRequest : requests) {
+            try {
+                stageRequest.activate().get();
+                stageRequestQueue.add(stageRequest);
+                _log.debug("Added {} to stageRequestQueue", stageRequest.getFileAttributes().getPnfsId());
+            } catch (ExecutionException | InterruptedException e) {
+                Throwable t = Throwables.getRootCause(e);
+                _log.error("Failed to activate request {}", t.getMessage());
+                stageRequest.failed(e);
+            }
+        }
+        _log.debug("Length of stageRequestQueue: {}", stageRequestQueue.size());
+    }
+
+    private Checksum calculateAdler32(File file) {
+        Adler32 newChecksum = new Adler32();
+        try {
+            byte [] fileArray = FileUtils.readFileToByteArray(file);
+            newChecksum.engineUpdate(fileArray, 0, fileArray.length);
+        } catch (IOException e) {
+            _log.error("Could not calculate checksum for file {}: ", file.getPath(), e);
+            return null;
+        }
+        return new Checksum(ChecksumType.ADLER32, newChecksum.engineDigest());
+    }
+
+    private Checksum calculateMd5(File file) { // TODO Needs testing
+        MessageDigest md = null;
+        try {
+            md = MessageDigest.getInstance("MD5");
+            InputStream in = new DataInputStream(new FileInputStream(file));
+            DigestInputStream din = new DigestInputStream(in, md);
+            _log.debug("Value of DigestInputStream MD5: {}", din.read());
```

Comment:
`din.read()` reads a byte. you probably need just `log.debug("Value of DigestInputStream MD5: {}", din);`